pkg/httpproxy/proxy.go

```diff
@@ -229,27 +234,96 @@ func (p *proxy) secretGetter(req *http.Request, cAuth string) SecretGetter {
 	}
 }
 
-func (p *proxy) checkCluster(req *http.Request, user user.Info, clusterID, credID string) (authorizer.Decision, error) {
-	cluster, err := p.clusters.Controller().Lister().Get("", clusterID)
-	if err != nil {
+// checkIndirectAccessViaCluster checks if the user has access to the cloud credential via being owner of a cluster associated to the cloud credential.
+// Currently, only EKS and provisioningv2 clusters are supported because those clusters have a cloud credential associated to them.
+// GKE and AKS clusters also have cloud credential associated to them, but those are checked via specific proxies (not the meta proxy).
+func (p *proxy) checkIndirectAccessViaCluster(req *http.Request, user user.Info, clusterID, credID string) (authorizer.Decision, error) {
+	var (
+		mgmtClusters []*v3.Cluster
+		provClusters []*prov.Cluster
+		err          error
+
+		i int
+		c interface{}
+	)
+	if clusterID == "" {
+		// If no clusterID is passed, then we check all clusters that the user has access to and are associated to the cloud credential.
+		// Both management and provisioning clusters should be checked.
+		mgmtClusters, err = p.mgmtClustersCache.GetByIndex(cluster.ByCloudCredential, credID)
+		if err != nil {
+			return authorizer.DecisionDeny, err
+		}
+
+		provClusters, err = p.provClustersCache.GetByIndex(provcluster.ByCloudCred, credID)
+	} else {
+		var c *v3.Cluster
+		c, err = p.mgmtClustersCache.Get(clusterID)
+		mgmtClusters = []*v3.Cluster{c}
+	}
+	if err != nil || len(mgmtClusters)+len(provClusters) == 0 {
 		return authorizer.DecisionDeny, err
 	}
-	if cluster.Spec.EKSConfig == nil {
-		return authorizer.DecisionDeny, nil
+
+	c = nextCluster(mgmtClusters, provClusters, &i)
+	for c != nil {
+		switch cluster := c.(type) {
+		case *v3.Cluster:
+			administratedByProvV2Cluster := imported.IsAdministratedByProvisioningCluster(cluster)
+			if (cluster.Spec.EKSConfig == nil || cluster.Spec.EKSConfig.AmazonCredentialSecret != credID) && !administratedByProvV2Cluster {
+				continue
+			}
+
+			if administratedByProvV2Cluster {
+				c, err = p.provClustersCache.GetByIndex(provcluster.ByCluster, cluster.Name)
+				if err != nil {
+					c = nextCluster(mgmtClusters, provClusters, &i)
+				}
+				// If there is no error, then we want to go through the loop again with the provisioning cluster.
+				continue
+			}
+
+			clusterID = cluster.Name
+
+		case *prov.Cluster:
+			if cluster.Spec.CloudCredentialSecretName != credID {
+				c = nextCluster(mgmtClusters, provClusters, &i)
+				continue
+			}
+
+			clusterID = cluster.Status.ClusterName
+		}
+
+		decision, _, err := p.authorizer.Authorize(req.Context(), authorizer.AttributesRecord{
+			User:            user,
+			Verb:            "update",
+			APIGroup:        v3.GroupName,
```

Comment:
Isn't this only for management clusters? Do we also need to check authorization on the provisioning cluster?

Reply:
If a user has access to a provisioning cluster, then they will also have access to the corresponding management cluster. So, in the case of a provisioning cluster, we use the `cluster.Status.ClusterName` which is the name of the management cluster.